Centralizes images into agnhost (part 1)

claudiubelu opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**

/kind feature

/sig testing

**What this PR does / why we need it**:

Refactors the functions used in agnhost into different modules, based on their functionality, leaving only the main in the base folder.

Future commits will add several functionalities to agnhost, so this change will be necessary to keep it clean.

Centralizes the following images into agnhost:

- fakegitserver
- hostexec
- liveness
- logs-generator
- no-snat-test
- no-snat-test-proxy
- port-forward-tester

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Related: #76342

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
NONE
```

claudiubelu committed:
test images: Refactors agnhost image

Refactors the functions used in agnhost into different modules,
based on their functionality, leaving only the main in the base
folder.

Future commits will add several functionalities to agnhost, so
this change will be necessary to keep it clean.

claudiubelu committed:
test images: Centralizes images into agnhost (part 1)

Centralizes the following images into agnhost:

- fakegitserver
- hostexec
- liveness
- logs-generator
- no-snat-test
- no-snat-test-proxy
- port-forward-tester

bclau commented:
/test pull-kubernetes-kubemark-e2e-gce-big

bclau commented:
/test pull-kubernetes-kubemark-e2e-gce-big

listx commented:
Looks like you are consolidating the separate Go modules into 1, just as this PR says. Sounds good to me.

/lgtm

@mkumatag Do you have any additional thoughts on this?

@bclau Should the people in #76342 also take a few minutes to review this PR? I ask because I only have a high-level understanding of what you are doing here and can't speak for the code changes.

bclau commented:
> 
> 
> Looks like you are consolidating the separate Go modules into 1, just as this PR says. Sounds good to me.
> 
> /lgtm
> 
> @mkumatag Do you have any additional thoughts on this?
> 
> @bclau Should the people in #76342 also take a few minutes to review this PR? I ask because I only have a high-level understanding of what you are doing here and can't speak for the code changes.

Any feedback is welcome, of course. As for the changes I've introduced in this PR, those are the most significant:

- I've mostly used ``git mv``, so the actual changes are easier to observe.
- I've removed ``go_binary`` from the ``BUILD`` files, since we're centralizing everything into the ``agnhost`` binary. We could leave it in, and make each be compilable on its own, if needed, but I don't know if we want that anymore.
- "refactored" the various flags into cobra flags, so they show up under their subcommand's help (e.g.: ``./agnhost help logs-generator``)
- moved any ``README``s to the ``agnhost``'s ``README`` and created new sections for the ones that didn't have any. I did my best to make them as helpful as possible, but if I've missed any important information, please let me know.

listx commented:
> > Looks like you are consolidating the separate Go modules into 1, just as this PR says. Sounds good to me.
> > /lgtm
> > @mkumatag Do you have any additional thoughts on this?
> > @bclau Should the people in #76342 also take a few minutes to review this PR? I ask because I only have a high-level understanding of what you are doing here and can't speak for the code changes.
> 
> Any feedback is welcome, of course. As for the changes I've introduced in this PR, those are the most significant:
> 
> * I've mostly used `git mv`, so the actual changes are easier to observe.
> * I've removed `go_binary` from the `BUILD` files, since we're centralizing everything into the `agnhost` binary. We could leave it in, and make each be compilable on its own, if needed, but I don't know if we want that anymore.
> * "refactored" the various flags into cobra flags, so they show up under their subcommand's help (e.g.: `./agnhost help logs-generator`)
> * moved any `README`s to the `agnhost`'s `README` and created new sections for the ones that didn't have any. I did my best to make them as helpful as possible, but if I've missed any important information, please let me know.
> 
> As far as testing goes, I've already built `claudiubelu/agnhost:1.1` with all the images enumerated under part 1 and part 2 (and a few others) centralized into it, and replaced those images with `agnhost` in tests, and as far as Conformance tests go, they've passed for me.

These changes sound fine to me, but I'd like to ask for others' opinions as well if possible before I approve as an owner.

Also, given the amount of activity you have in this area, it might be worth trying to add you as an approver (even though you would still need others' approvals in PRs). Are you a kubernetes org member yet?

bclau commented:
> 
> These changes sound fine to me, but I'd like to ask for others' opinions as well if possible before I approve as an owner.
> 
> Also, given the amount of activity you have in this area, it might be worth trying to add you as an approver (even though you would still need others' approvals in PRs). Are you a kubernetes org member yet?

I am: https://github.com/kubernetes/org/issues/509

listx commented:
@mkumatag @luxas Thoughts?

bclau commented:
/assign @timothysc

mkumatag commented on test/images/agnhost/VERSION:
```diff
@@ -1 +1 @@
-1.0
+1.1
```
Let us bump up to the major release because a lot of changes went in(something like 2.0)

bclau replied:
Done

mkumatag commented on test/images/agnhost/README.md:
```diff
@@ -17,18 +17,8 @@ cases for the same tested behaviour.
-- `help`: Prints the binary's help menu. Additionally, it can be followed by another
-  subcommand in order to get more information about that subcommand, including its
-  possible arguments.
+The `agnhost` binary has several subcommands which are can be used to test different
```
I'm not really sure what this `agnhost` means but question is - is this name still relevant after adding so many other subcommands?

bclau replied:
Eh, less so. So, the name ``agnhost`` comes from "agnhost host", which basically means that it'll work the same on any OS. But it is still true, the image still contains the common Linux utilities on both Linux and Windows, including /bin/sh, nc, etc., and 3 of the current subcommands still accomplish the original goal. Adding the other subcommands doesn't necessarely go against its goal.

mkumatag commented:
> @mkumatag @luxas Thoughts?

overall changes looks okay to me, lets do the following thing to make sure new changes works fine!

@listx push the new image 
@bclau let us submit the commit with using above image in manifest.go and tests once @listx pushes the image to make sure all the conformance passes.

bclau commented:
> 
> 
> > @mkumatag @luxas Thoughts?
> 
> overall changes looks okay to me, lets do the following thing to make sure new changes works fine!
> 
> @listx push the new image
> @bclau let us submit the commit with using above image in manifest.go and tests once @listx pushes the image to make sure all the conformance passes.

Already sent the PR for that: https://github.com/kubernetes/kubernetes/pull/78394

mkumatag commented:
> > > @mkumatag @luxas Thoughts?
> > 
> > 
> > overall changes looks okay to me, lets do the following thing to make sure new changes works fine!
> > @listx push the new image
> > @bclau let us submit the commit with using above image in manifest.go and tests once @listx pushes the image to make sure all the conformance passes.
> 
> Already sent the PR for that: #78394

Thanks @bclau, @listx can you build and push new `agnhost` image to pass other PR?

claudiubelu committed:
Updates agnhost README and Dockerfile files

logs-generator used to be configured through env variables, but with
the image centralization, that changed to flags.

This commit updates the agnhost README to reflect that.

listx commented:
@mkumatag I'll try to get the images pushed later today, caught up with something at the moment.

Sorry about the delay!

listx commented:
I pushed the images. Log below:

```
./image-util.sh push agnhost
The push refers to repository [gcr.io/kubernetes-e2e-test-images/agnhost-amd64]
5e365e26c3ca: Pushed
3aa628b98439: Pushed
721384ec99e5: Layer already exists
2.0: digest: sha256:2c0b5d8e1da6bb525f0d4b51c220dac7e8d4081edab7ef728e7931827390eabc size: 950                                                                                    
The push refers to repository [gcr.io/kubernetes-e2e-test-images/agnhost-arm]
a88e19706e8a: Pushed
bfa597a7eed6: Pushed
350c5d4fce68: Pushed
845b5c25f81e: Layer already exists
2.0: digest: sha256:fdd877012a1c31ff6cc6728cf30911bb77cb7119a127af86249deaaf63a6aadc size: 1161                                                                                   
The push refers to repository [gcr.io/kubernetes-e2e-test-images/agnhost-arm64]
5cc225da45e7: Pushed
15901ef31942: Pushed
874c57655f97: Pushed
56e86ee0b452: Layer already exists
2.0: digest: sha256:6ac5972b281c12dbb9aca99b8e498de443539be0ccc7fb0f470cd166ff936d28 size: 1161                                                                                   
The push refers to repository [gcr.io/kubernetes-e2e-test-images/agnhost-ppc64le]
2396d6e23399: Pushed
109e2b3d6de0: Pushed
796f1353ab3d: Pushed
85e899b8f0cc: Layer already exists
2.0: digest: sha256:1c58fa90acc5fd9e5577cde4564de3c1d2426daf6cc3c0d95cd2a3182fd547e8 size: 1161                                                                                   
The push refers to repository [gcr.io/kubernetes-e2e-test-images/agnhost-s390x]
021229db5da7: Pushed
16bc4535b331: Pushed
f129ff7d977b: Layer already exists
2.0: digest: sha256:ac24934dcaf3e0d274fefe18abc69c5f91d074b2fd672f44f18523e0d130ac26 size: 950                                                                                    
Created manifest list gcr.io/kubernetes-e2e-test-images/agnhost:2.0
Pushed ref gcr.io/kubernetes-e2e-test-images/agnhost@sha256:2c0b5d8e1da6bb525f0d4b51c220dac7e8d4081edab7ef728e7931827390eabc with digest: sha256:2c0b5d8e1da6bb525f0d4b51c220dac7e8d4081edab7ef728e7931827390eabc
Pushed ref gcr.io/kubernetes-e2e-test-images/agnhost@sha256:fdd877012a1c31ff6cc6728cf30911bb77cb7119a127af86249deaaf63a6aadc with digest: sha256:fdd877012a1c31ff6cc6728cf30911bb77cb7119a127af86249deaaf63a6aadc
Pushed ref gcr.io/kubernetes-e2e-test-images/agnhost@sha256:6ac5972b281c12dbb9aca99b8e498de443539be0ccc7fb0f470cd166ff936d28 with digest: sha256:6ac5972b281c12dbb9aca99b8e498de443539be0ccc7fb0f470cd166ff936d28
Pushed ref gcr.io/kubernetes-e2e-test-images/agnhost@sha256:1c58fa90acc5fd9e5577cde4564de3c1d2426daf6cc3c0d95cd2a3182fd547e8 with digest: sha256:1c58fa90acc5fd9e5577cde4564de3c1d2426daf6cc3c0d95cd2a3182fd547e8
Pushed ref gcr.io/kubernetes-e2e-test-images/agnhost@sha256:ac24934dcaf3e0d274fefe18abc69c5f91d074b2fd672f44f18523e0d130ac26 with digest: sha256:ac24934dcaf3e0d274fefe18abc69c5f91d074b2fd672f44f18523e0d130ac26
sha256:75990e17b63596dc69c3fd3161a281445048457da1555caebfecd5c1215a7c2e
```

bclau commented:
> 
> 
> I pushed the images. Log below:
> 

Thanks! It seems that the tests have passed: https://github.com/kubernetes/kubernetes/pull/78389

timothysc commented:
/assign @dims 

Who on the google side do we need to loop in here?

mkumatag commented:
> /assign @dims
> 
> Who on the google side do we need to loop in here?

@listx already working on this.

mkumatag commented:
I see all green here - https://github.com/kubernetes/kubernetes/pull/78394, I think we are good.

/lgtm

mkumatag commented:
/approve

spiffxp commented on test/images/agnhost/README.md:
```diff
@@ -72,8 +62,213 @@ created with the `pause` argument instead, allowing us execute multiple commands
 
-The image can be found at `gcr.io/kubernetes-e2e-test-images/agnhost:1.0` for Linux
-containers, and `e2eteam/agnhost:1.0` for Windows containers. In the future, the same
+The image can be found at `gcr.io/kubernetes-e2e-test-images/agnhost:1.1` for Linux
```
```suggestion
The image can be found at `gcr.io/kubernetes-e2e-test-images/agnhost:2.0` for Linux
```

bclau replied:
Oups, I forgot to update the version in the documentation.

spiffxp commented on test/images/agnhost/README.md:
```diff
@@ -72,8 +62,213 @@ created with the `pause` argument instead, allowing us execute multiple commands
-The image can be found at `gcr.io/kubernetes-e2e-test-images/agnhost:1.0` for Linux
-containers, and `e2eteam/agnhost:1.0` for Windows containers. In the future, the same
+The image can be found at `gcr.io/kubernetes-e2e-test-images/agnhost:1.1` for Linux
+containers, and `e2eteam/agnhost:1.1` for Windows containers. In the future, the same
```
```suggestion
containers, and `e2eteam/agnhost:2.0` for Windows containers. In the future, the same
```

bclau replied:
ditto.

spiffxp commented on test/images/agnhost/Dockerfile:
```diff
@@ -12,7 +12,20 @@
+
+# from logs-generator
+ENV LOGS_GENERATOR_LINES_TOTAL 1
+ENV LOGS_GENERATOR_DURATION 1s
```
non-blocking nit: I can't tell that these are actually consumed by `log_generator.go` so it's unclear why they're necessary

bclau replied:
True, it seems that the env vars are not consumed, the logs-generator binary was expecting those params as flags from the initial commit, yet the documentation was also talking about env vars. The tests also sets those env vars.

spiffxp commented on test/images/agnhost/README.md:
```diff
@@ -72,8 +62,213 @@ created with the `pause` argument instead, allowing us execute multiple commands
+Parameters are passed through environment variables. There are no defaults, you should always
+set up container parameters. Total number of line is parametrized through env variable
+`LOGS_GENERATOR_LINES_TOTAL` and duration in go format is parametrized through env variable
+`LOGS_GENERATOR_DURATION`.
```
this paragraph is incorrect: near as I can tell, `logs_generator.go` only takes flags, not env vars, and I can't tell that agnhost plumbs the env vars through?

bclau replied:
This part of the documentation was pretty much copied from the original documentation logs-generator had. Interestingly, it was like this from the beggining: accepting those parameters as flags and not as env vars, yet the documentation is talking about env vars. The tests also sets those env vars. In the replacement PR, I've replaced that with actual flags: https://github.com/kubernetes/kubernetes/pull/78394/files#diff-7dc1ebd3b2f2f6ca1ff248e6601016a6R108

spiffxp commented on test/images/logs-generator/Dockerfile:
```diff
@@ -1,22 +0,0 @@
-
-COPY logs-generator /
-
-CMD ["sh", "-c", "/logs-generator --logtostderr --log-lines-total=${LOGS_GENERATOR_LINES_TOTAL} --run-duration=${LOGS_GENERATOR_DURATION}"]
```
this explains my confusion: the old logs-generator container uses env vars to pass to the parameters; the agnhost refactor does not

johnSchnake replied:
Yea the new one just uses flags. Based on https://github.com/kubernetes/kubernetes/pull/78394/files#diff-7dc1ebd3b2f2f6ca1ff248e6601016a6R108 it looks like the invocation was updated to use the flags instead of env vars too, so the doc notes about the env vars can be removed.

bclau replied:
Good catch, there's no longer a need for those env variables. I'll also remove them from the Dockerfile.

spiffxp reviewed (changes requested):
/lgtm cancel
this is really close I just have some tiny nits

bclau commented:
I've updated the README with the new usage and removed the env vars from the Dockerfile.

bclau commented:
/test pull-kubernetes-e2e-gce

spiffxp reviewed (approved):
/lgtm
thanks for the updates

johnSchnake commented on test/images/agnhost/logs-generator/logs_generator.go:
```diff
@@ -38,27 +39,35 @@ var (
+	Use:   "logs-generator",
+	Short: "Outputs lines of logs to stdout uniformly",
+	Long:  "Outputs <linesTotal> lines of logs to stdout uniformly for <duration>",
+	Args:  cobra.MaximumNArgs(2),
```
Shouldnt this be 0 args; the variables are given via flags not arguments.

bclau replied:
Oups, you're right. Then that means that pretty much all commands should have this at 0.

claudiubelu committed:
tests: Sets MaximumNArgs for agnhost subcommands to 0

The subcommands rely on flags, not args. Because of this, the
cobra.MaximumNArgs should be set 0.

claudiubelu committed:
Updates agnhost image version in documentation

The version was bumped to 2.1, not 1.1.

Co-Authored-By: Aaron Crickenberger <spiffxp@google.com>

spiffxp commented:
/lgtm
